### ISSUE TITLE
Panic while running binary

### ISSUE THREAD

**Issue by alexkohler** [nikola-kocic/i3-switch-tabs#2]
I'm running into the following while running the `i3-switch-tabs` binary:

```
$ ./i3-switch-tabs left
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /checkout/src/libcore/option.rs:335:20
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Am I missing something? Thanks for writing this nifty lil tool :)

**Comment by nikola-kocic** [comment]
Please run `$ RUST_BACKTRACE=1 ./i3-switch-tabs left`. I will soon release new version that will print an error when it fails.
Note that this tool will only work if you are running [i3wm](https://i3wm.org/).

**Comment by alexkohler** [comment]
Sure. I am running `i3`, I seem to get the same error regardless of whether I have a tabbed layout or not.
```
$ RUST_BACKTRACE=1 ./i3-switch-tabs left
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /checkout/src/libcore/option.rs:335:20
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:397
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:611
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:572
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:522
   7: rust_begin_unwind
             at /checkout/src/libstd/panicking.rs:498
   8: core::panicking::panic_fmt
             at /checkout/src/libcore/panicking.rs:71
   9: core::panicking::panic
             at /checkout/src/libcore/panicking.rs:51
  10: i3ipc::common::build_tree
  11: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::spec_extend
  12: i3ipc::common::build_tree
  13: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::spec_extend
  14: i3ipc::common::build_tree
  15: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::spec_extend
  16: i3ipc::common::build_tree
  17: i3ipc::I3Connection::get_tree
  18: i3_switch_tabs::main
  19: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:99
  20: std::rt::lang_start
             at /checkout/src/libstd/panicking.rs:459
             at /checkout/src/libstd/panic.rs:361
             at /checkout/src/libstd/rt.rs:61
  21: __libc_start_main
  22: _start
```

**Comment by nikola-kocic** [comment]
This seems to be an issue with code in dependency (i3ipc). I created new release (0.1.2) with updated i3ipc, try it out.

If you get crash again, try downloading debug build, and running it same as above (`$ RUST_BACKTRACE=1 ./i3-switch-tabs left`), it should give me more info.

Also, you can try running `$ i3-msg -t get_tree`, to see if that works fine.
If it does, and if it does not contain any sensitive data (it lists the running programs and their titles), you can send the output of that command to this ticket or to my e-mail, so I can debug the issue.

**Comment by alexkohler** [comment]
No luck with 0.1.2, here's the output:

```
$ RUST_BACKTRACE=1 ./i3-switch-tabs left
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', libcore/option.rs:345:21
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:477
   5: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:391
   6: rust_begin_unwind
             at libstd/panicking.rs:326
   7: core::panicking::panic_fmt
             at libcore/panicking.rs:77
   8: core::panicking::panic
             at libcore/panicking.rs:52
   9: <core::option::Option<T>>::unwrap
             at libcore/macros.rs:20
  10: i3ipc::common::build_tree
             at /home/nikola/.cargo/registry/src/github.com-1ecc6299db9ec823/i3ipc-0.9.0/src/common.rs:41
  11: i3ipc::common::build_tree::{{closure}}
             at /home/nikola/.cargo/registry/src/github.com-1ecc6299db9ec823/i3ipc-0.9.0/src/common.rs:21
  12: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &'a mut F>::call_once
             at libcore/ops/function.rs:286
  13: <core::option::Option<T>>::map
             at libcore/option.rs:414
  14: <core::iter::Map<I, F> as core::iter::iterator::Iterator>::next
             at libcore/iter/mod.rs:1394
  15: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::spec_extend
             at liballoc/vec.rs:1903
  16: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::from_iter
             at liballoc/vec.rs:1886
  17: <alloc::vec::Vec<T> as core::iter::traits::FromIterator<T>>::from_iter
             at liballoc/vec.rs:1772
  18: core::iter::iterator::Iterator::collect
             at libcore/iter/iterator.rs:1415
  19: i3ipc::common::build_tree
             at /home/nikola/.cargo/registry/src/github.com-1ecc6299db9ec823/i3ipc-0.9.0/src/common.rs:18
  20: i3ipc::common::build_tree::{{closure}}
             at /home/nikola/.cargo/registry/src/github.com-1ecc6299db9ec823/i3ipc-0.9.0/src/common.rs:21
  21: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &'a mut F>::call_once
             at libcore/ops/function.rs:286
  22: <core::option::Option<T>>::map
             at libcore/option.rs:414
  23: <core::iter::Map<I, F> as core::iter::iterator::Iterator>::next
             at libcore/iter/mod.rs:1394
  24: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::spec_extend
             at liballoc/vec.rs:1903
  25: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::from_iter
             at liballoc/vec.rs:1886
  26: <alloc::vec::Vec<T> as core::iter::traits::FromIterator<T>>::from_iter
             at liballoc/vec.rs:1772
  27: core::iter::iterator::Iterator::collect
             at libcore/iter/iterator.rs:1415
  28: i3ipc::common::build_tree
             at /home/nikola/.cargo/registry/src/github.com-1ecc6299db9ec823/i3ipc-0.9.0/src/common.rs:18
  29: i3ipc::common::build_tree::{{closure}}
             at /home/nikola/.cargo/registry/src/github.com-1ecc6299db9ec823/i3ipc-0.9.0/src/common.rs:21
  30: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &'a mut F>::call_once
             at libcore/ops/function.rs:286
  31: <core::option::Option<T>>::map
             at libcore/option.rs:414
  32: <core::iter::Map<I, F> as core::iter::iterator::Iterator>::next
             at libcore/iter/mod.rs:1394
  33: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::spec_extend
             at liballoc/vec.rs:1903
  34: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::from_iter
             at liballoc/vec.rs:1886
  35: <alloc::vec::Vec<T> as core::iter::traits::FromIterator<T>>::from_iter
             at liballoc/vec.rs:1772
  36: core::iter::iterator::Iterator::collect
             at libcore/iter/iterator.rs:1415
  37: i3ipc::common::build_tree
             at /home/nikola/.cargo/registry/src/github.com-1ecc6299db9ec823/i3ipc-0.9.0/src/common.rs:18
  38: i3ipc::I3Connection::get_tree
             at /home/nikola/.cargo/registry/src/github.com-1ecc6299db9ec823/i3ipc-0.9.0/src/lib.rs:386
  39: i3_switch_tabs::superfocus
             at src/main.rs:51
  40: i3_switch_tabs::main
             at src/main.rs:79
  41: std::rt::lang_start::{{closure}}
             at libstd/rt.rs:74
  42: std::panicking::try::do_call
             at libstd/rt.rs:59
             at libstd/panicking.rs:310
  43: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:103
  44: std::rt::lang_start_internal
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at libstd/rt.rs:58
  45: std::rt::lang_start
             at libstd/rt.rs:74
  46: main
  47: __libc_start_main
  48: _start
```

I'll send the get_tree output via email - what is your email?

**Comment by nikola-kocic** [comment]
Thanks, this stacktrace means that either "type" property of some tree element is missing (unlikely), or that it's not of string type.

Which version of i3wm are you running? It seems that "type" property became string in 4.8 (released on Jun 15, 2014), but Ubuntu 14.04 for example has version 4.7.2.

I released version 0.1.3 which checks the version of i3wm and added that note to README.

This is exact commit in i3wm that changes "type" property value type: https://github.com/i3/i3/commit/6800524f

**Comment by alexkohler** [comment]
Ah, yes - I am on 4.7. Thanks a ton for all the help!
